src/Queue/DoctrineQueue.php

```
@@ -182,19 +225,162 @@ public function pop(array $options = []): ?JobInterface
+                return null;
+            }
+
+            if ($rowCount > 1) {
+                // That should not be possible, but what if ...? Reset them...
+                $queryBuilder = $this->connection->createQueryBuilder();
+                $queryBuilder
+                    ->update($this->options->getTableName(), 'job')
+
+                    ->set('job.status = :newStatus')
+                    ->setParameter('newStatus', static::STATUS_PENDING)
+
+                    ->set('job.workerId = NULL')
+
+                    ->where('job.status = :oldStatus')
+                    ->setParameter('oldStatus', static::STATUS_ALLOCATING)
+
+                    ->andWhere('job.workerId = :workerId')
+                    ->setParameter('workerId', $this->workerId)
+
+                    ->andWhere('job.queue = :queue')
+                    ->setParameter('queue', $this->getName());
+
+                $this->connection->executeQuery(
+                    $queryBuilder->getSQL(),
+                    $queryBuilder->getParameters(),
+                    $queryBuilder->getParameterTypes()
+                );
+
+                throw new LogicException(
+                    'More than 1 job was assigned to our worker despite using UPDATE with LIMIT 1!'
+                );
+            }
+
+            $queryBuilder = $this->connection->createQueryBuilder();
+            $queryBuilder
+                ->select('*')
+                ->from($this->options->getTableName())
+
+                ->where('status = :oldStatus')
+                ->setParameter('oldStatus', static::STATUS_ALLOCATING)
+
+                ->andWhere('workerId = :workerId')
+                ->setParameter('workerId', $this->workerId);
+
+            $result = $this->connection->executeQuery(
+                $queryBuilder->getSQL(),
+                $queryBuilder->getParameters(),
+                $queryBuilder->getParameterTypes()
+            );
+
+            if ($row = $result->fetch()) {
```

You could also throw a logic exception here if we get more than one result. That is like really unlikely but could happen if there is a job from a previous worker that got allocated, but was never run.

---

That shouldnt happen because the worker ID should be unique (PID + a random int). Anyway, throwing an exception!